File: object_relations/app/models/customer.rb
class Customer

  attr_accessor :name
  @@all = []

  def initialize(name)
    @name = name
    @@all << self
  end  

  def self.all
    @@all
  end
  
  def self.find_by_name(name)
    self.all.find { |customer| customer.name == name }
  end

  def add_review(content, restaurant_name)    
    restaurant = Restaurant.find_or_create_restaurant(restaurant_name)
    Review.new(content, restaurant, self) 
  end

  def reviews
    Review.all.select { |review| review.customer == self }
  end

  def add_restaurant(restaurant_name)
    restaurant = Restaurant.find_by_name(restaurant_name)
      if !restaurant
        Restaurant.new(restaurant_name)
      end
  end

  def restaurants
    self.reviews.map { |review| review.restaurant }
  end

end

# customer = customer.add_review('it was good', Restaurant.new)
# a new review tied ot the restaurant
# and tied to the customer

  # Customer#add_review
  # Customer#add_restaurant

File: javascript/models/customer.js
class Customer {
  constructor(name) {
    this.name = name
    store.customers.push(this)
  }

  all() {
    return store.customers
  }

  find_by_name(name) {
    for (let i = 0; i < store.customers.length; i++) {
      if (store.customers[i].name === name) {
        return store.customers[i]
      }
    }
  }

  add_review(content, restaurant_name) {
    let restaurant = Restaurant.prototype.find_or_create_restaurant(restaurant_name)
    new Review(content, restaurant, this)  
  }

  add_restaurant(name) {
    new Restaurant(name)
  }
}


